### PR TITLE
Add RTE keyboard navigation in editing

### PR DESCRIPTION
**Pull request by florianduros**
## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Fix https://github.com/vector-im/element-web/issues/23621

- Add keyboard navigation (arrow up and arrow down) to circle across the messages in the timeline
- Use the same keyboard action handling than in the current `MessageComposer`
- Currently, only the RTE in Rich Text mode is available in editing mode. The Plain text mode isn't in the scope of this PR.
- When arrow down is pressed on the last message, the editing mode is closed
- I wrote cypress tests but they are in failure in the CI (working in local), I removed them.

https://user-images.githubusercontent.com/2621378/215782335-e70a69d6-e91c-46ec-83fa-611fcf3bea78.mov

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add RTE keyboard navigation in editing ([\#9980](https://github.com/matrix-org/matrix-react-sdk/pull/9980)). Fixes vector-im/element-web#23621. Contributed by @florianduros.

### CODE REVIEW

File: src/components/views/rooms/wysiwyg_composer/hooks/useInputEventProcessor.ts [hunk 1]
```diff
@@ -14,40 +14,164 @@ See the License for the specific language governing permissions and
...
-
-export function useInputEventProcessor(onSend: () => void): (event: WysiwygEvent) => WysiwygEvent | null {
+export function useInputEventProcessor(
+    onSend: () => void,
+    initialContent?: string,
+): (event: WysiwygEvent, composer: Wysiwyg, editor: HTMLElement) => WysiwygEvent | null {
+    const roomContext = useRoomContext();
+    const composerContext = useComposerContext();
+    const mxClient = useMatrixClientContext();
     const isCtrlEnter = useSettingValue<boolean>("MessageComposerInput.ctrlEnterToSend");
```
Comment (alunturner) [inline src/components/views/rooms/wysiwyg_composer/hooks/useInputEventProcessor.ts:43]: Maybe picky, but could we rename this to `isCtrlEnterToSend`? I know it's a bit longer but I think the clarity is helpful, I was reading this as meaning something like "the user has pressed control and enter at the same time" whereas it actually means "the user has got the setting ctrl enter to send turned on".

File: src/components/views/rooms/wysiwyg_composer/hooks/useInputEventProcessor.ts [hunk 2]
```diff
@@ -14,40 +14,164 @@ See the License for the specific language governing permissions and
...
+            }
+
+            const isDispatched = dispatchEditEvent(event, true, editorStateTransfer, roomContext, mxClient);
+            if (!isDispatched) {
+                endEditing(roomContext);
+                event.preventDefault();
+                event.stopPropagation();
+            }
+
+            break;
```
Comment (florianduros) [inline src/components/views/rooms/wysiwyg_composer/hooks/useInputEventProcessor.ts:123]: Should return null here

File: test/components/views/rooms/wysiwyg_composer/EditWysiwygComposer-test.tsx [hunk 3]
```diff
@@ -318,4 +342,290 @@ describe("EditWysiwygComposer", () => {
...
+        describe("Moving up", () => {
+            it("Should not moving when caret is not at beginning of the text", async () => {
+                // When
+                const { textbox, spyDispatcher } = await setup();
+                const textNode = textbox.firstChild;
+                await select({
+                    anchorNode: textNode,
+                    anchorOffset: 1,
+                    focusNode: textNode,
+                    focusOffset: 2,
```
Comment (alunturner) [inline test/components/views/rooms/wysiwyg_composer/EditWysiwygComposer-test.tsx:380]: I may be misreading this, but I think if we have both `anchorNode` and `focusNode` pointing at the same node, but different offsets, we don't have a caret in this test do we? I thought for a caret both offsets would have to be the same

Comment (florianduros) [reply test/components/views/rooms/wysiwyg_composer/EditWysiwygComposer-test.tsx:380]: Here, there is a selection starting from the anchorOffset(0) to the focusOffset(1). The caret is on the focusOffset

**Review by alunturner — COMMENTED** [review]
I like the tests for `EditWysiwygComposer`. On top of the comments, there's one minor presentation issue that I've found when playing for it and that's that, when editing a message, bullet points become discs as opposed to circles (ie they are unfilled shapes). Would we be able to get those to be consistent with the timeline display?

**Comment by florianduros** [comment]
> I like the tests for `EditWysiwygComposer`. On top of the comments, there's one minor presentation issue that I've found when playing for it and that's that, when editing a message, bullet points become discs as opposed to circles (ie they are unfilled shapes). Would we be able to get those to be consistent with the timeline display?

Yes, I think it's possible to do it. We need to apply the same style between the editor and the timeline display. In the same way than the code blocks and inline code.

**Review by alunturner — APPROVED** [review]

**Pull request merged**